File: src/leetcode/SortList.java
package leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

public class SortList {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null){
            return head;
        }
        ListNode slow = head, fast = head, prev = null;
        while (fast  != null && fast.next != null){
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        if (prev != null)prev.next = null;
        ListNode first = sortList(head);
        ListNode second = sortList(slow);
        return merge(first, second);
    }
    
    public ListNode merge(ListNode first, ListNode second){
        ListNode head = null, node = null;
        while( first != null || second != null){
            int v1 = Integer.MAX_VALUE;
            int v2 = Integer.MAX_VALUE;
            if (first != null)v1 = first.val;
            if (second != null) v2 = second.val;
            if( v1 < v2){
                if(head == null){
                    head = first;
                    node = first;
                }else{
                    node.next = first;
                    node = node.next;
                }
                first = first.next;
            }else{
                if(head == null){
                    head = second;
                    node = second;
                }else{
                    node.next = second;
                    node = node.next;
                }
                second = second.next;
            }
        }
        return head;
    }
}

File: python/leetcode/SortList.py
class SortList:
    def sortList(self, head):
        if head is None or head.next is None:
            return head
        slow = head
        fast = head
        prev = None
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next
            if fast:
                fast = fast.next
        prev.next = None
        l = self.sortList(head)
        r = self.sortList(slow)
        newHead = None
        node = None
        while l and r:
            if l.val < r.val:
                if newHead is None:
                    newHead = l
                else:
                    node.next = l
                node = l
                l = l.next
            else:
                if newHead is None:
                    newHead = r
                else:
                    node.next = r
                node = r
                r = r.next
        if l:
            node.next = l
        if r:
            node.next = r
        return newHead